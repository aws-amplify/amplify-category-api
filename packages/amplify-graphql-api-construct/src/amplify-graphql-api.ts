import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as cfninclude from 'aws-cdk-lib/cloudformation-include';
import * as appsync from 'aws-cdk-lib/aws-appsync';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';
import { executeTransform } from '@aws-amplify/graphql-transformer';
import { ResolverConfig } from '@aws-amplify/graphql-transformer-core';
import {
  convertAuthorizationModesToTransformerAuthConfig,
  getAuthParameters,
  rewriteAssets,
  preprocessGraphQLSchema,
  generateConstructExports,
  persistStackAssets,
} from './internal';
import type { AuthorizationMode, AmplifyGraphQlApiResources } from './types';
import { TransformerPluginProvider } from '../../amplify-graphql-transformer-interfaces';
import { NoopFeatureFlagProvider } from './internal/NoopFeatureFlagProvider';

export type AmplifyApiCompatibilityLayer = {
  stackMappings?: Record<string, string>;
  slotOverrides?: Record<string, string>;
  customTransformers?: TransformerPluginProvider[];
  predictionsBucket?: s3.IBucket;
};

export type AmplifyGraphQlApiProps = {
  schema: appsync.SchemaFile | string;
  envOverride?: string;
  apiName: string;
  authorizationModes?: AuthorizationMode[];
  resolverConfig?: ResolverConfig;
  _compat?: AmplifyApiCompatibilityLayer;
};

export class AmplifyGraphQlApi extends Construct {
  public readonly resources: AmplifyGraphQlApiResources;
  public readonly env: string;

  constructor(scope: Construct, id: string, props: AmplifyGraphQlApiProps) {
    super(scope, id);

    const {
      schema: modelSchema,
      authorizationModes,
      envOverride,
      _compat,
      resolverConfig,
    } = props;

    const assetDir = fs.mkdtempSync(path.join(os.tmpdir(), 'transformer'));

    const preprocessedSchema = preprocessGraphQLSchema(modelSchema);
    const { authConfig, identityPoolId, adminRoles } = convertAuthorizationModesToTransformerAuthConfig(authorizationModes);

    const {
      rootStack, stacks, resolvers, schema, functions,
    } = executeTransform({
      schema: preprocessedSchema,
      // resolvers: _compat?.slotOverrides ?? {},
      userDefinedSlots: {}, // TODO: Figure this out.
      stacks: {},
      transformersFactoryArgs: {
        authConfig,
        identityPoolId,
        adminRoles,
        customTransformers: _compat?.customTransformers ?? [],
        ...(_compat?.predictionsBucket ? { predictionsConfig: { bucketName: _compat.predictionsBucket.bucketName } } : {}),
      },
      authConfig,
      stackMapping: _compat?.stackMappings ?? {},
      resolverConfig,
      featureFlags: new NoopFeatureFlagProvider(), // TODO: Figure this out.
    });

    rewriteAssets(this, {
      assetDir,
      schema,
      resolvers,
      functions,
      rootStack,
      stacks,
    });

    const { rootTemplateFile, nestedStackConfig } = persistStackAssets({
      assetDir,
      rootStack,
      stacks,
    });

    // Allow env as an override prop, otherwise retrieve from context, and use value 'NONE' if no value can be found.
    // env is required for logical id suffixing, as well as Exports from the nested stacks.
    // Allow export so customers can reuse the env in their own references downstream.
    this.env = envOverride ?? this.node.tryGetContext('env') ?? 'NONE';
    if (this.env.length > 8) {
      throw new Error(`envOverride prop or cdk --context env must have a length <= 8, found ${this.env}`);
    }

    const transformerStack = new cfninclude.CfnInclude(this, 'RootStack', {
      templateFile: rootTemplateFile,
      loadNestedStacks: nestedStackConfig,
      parameters: {
        AppSyncApiName: props.apiName,
        env: this.env,
        S3DeploymentBucket: cdk.DefaultStackSynthesizer.DEFAULT_FILE_ASSETS_BUCKET_NAME,
        S3DeploymentRootKey: cdk.DefaultStackSynthesizer.DEFAULT_FILE_ASSET_KEY_ARN_EXPORT_NAME,
        ...getAuthParameters(authorizationModes),
      },
      preserveLogicalIds: true,
    });

    this.resources = generateConstructExports(rootStack, stacks, transformerStack);
  }
}
