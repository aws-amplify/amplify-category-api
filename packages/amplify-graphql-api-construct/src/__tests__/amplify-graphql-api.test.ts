import * as cdk from 'aws-cdk-lib';
import { AmplifyGraphqlApi } from '../amplify-graphql-api';

describe('AmplifyGraphqlApi', () => {
  describe('getGeneratedFunctionSlots', () => {
    it('returns slots, and includes a known slot', () => {
      const api = new AmplifyGraphqlApi(new cdk.Stack(), 'TestApi', {
        schema: /* GraphQL */ `
          type Todo @model @auth(rules: [{ allow: public }]) {
            description: String!
          }
        `,
        authorizationConfig: {
          apiKeyConfig: { expires: cdk.Duration.days(7) },
        },
      });
  
      const generatedFunctionSlots = api.getGeneratedFunctionSlots();
  
      expect(generatedFunctionSlots.length).toEqual(20);
      expect(generatedFunctionSlots).toEqual(expect.arrayContaining([
        expect.objectContaining({
          typeName: 'Mutation',
          fieldName: 'createTodo',
          slotName: 'postAuth',
          slotIndex: 1,
          function: expect.objectContaining({
            requestMappingTemplate: expect.any(String)
          })
        }),
      ]));
    });
  });
});
