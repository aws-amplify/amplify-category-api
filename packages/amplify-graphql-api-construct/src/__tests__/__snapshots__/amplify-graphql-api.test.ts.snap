// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmplifyGraphQlApi transforms a simple schema 1`] = `
Object {
  "Outputs": Object {
    "GraphQLAPIEndpointOutput": Object {
      "Description": "Your GraphQL API endpoint.",
      "Export": Object {
        "Name": Object {
          "Fn::Join": Array [
            ":",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "GraphQLApiEndpoint",
            ],
          ],
        },
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "GraphQLAPI",
          "GraphQLUrl",
        ],
      },
    },
    "GraphQLAPIIdOutput": Object {
      "Description": "Your GraphQL API ID.",
      "Export": Object {
        "Name": Object {
          "Fn::Join": Array [
            ":",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "GraphQLApiId",
            ],
          ],
        },
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "GraphQLAPI",
          "ApiId",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "DynamoDBBillingMode": Object {
      "AllowedValues": Array [
        "PAY_PER_REQUEST",
        "PROVISIONED",
      ],
      "Default": "PAY_PER_REQUEST",
      "Description": "Configure @model types to create DynamoDB tables with PAY_PER_REQUEST or PROVISIONED billing modes.",
      "Type": "String",
    },
    "DynamoDBEnablePointInTimeRecovery": Object {
      "AllowedValues": Array [
        "true",
        "false",
      ],
      "Default": "false",
      "Description": "Whether to enable Point in Time Recovery on the table.",
      "Type": "String",
    },
    "DynamoDBEnableServerSideEncryption": Object {
      "AllowedValues": Array [
        "true",
        "false",
      ],
      "Default": "true",
      "Description": "Enable server side encryption powered by KMS.",
      "Type": "String",
    },
    "DynamoDBModelTableReadIOPS": Object {
      "Default": 5,
      "Description": "The number of read IOPS the table should support.",
      "Type": "Number",
    },
    "DynamoDBModelTableWriteIOPS": Object {
      "Default": 5,
      "Description": "The number of write IOPS the table should support.",
      "Type": "Number",
    },
  },
  "Resources": Object {
    "GraphQLAPI": Object {
      "Properties": Object {
        "AuthenticationType": "AMAZON_COGNITO_USER_POOLS",
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              "TestApi",
              "-",
              "NONE",
            ],
          ],
        },
        "UserPoolConfig": Object {
          "AwsRegion": Object {
            "Ref": "AWS::Region",
          },
          "DefaultAction": "ALLOW",
          "UserPoolId": "ImportedUserPoolId",
        },
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "GraphQLAPINONEDS95A13CF0": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphQLAPI",
            "ApiId",
          ],
        },
        "Description": "None Data Source for Pipeline functions",
        "Name": "NONE_DS",
        "Type": "NONE",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "GraphQLAPITransformerSchema3CB2AE18": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphQLAPI",
            "ApiId",
          ],
        },
        "DefinitionS3Location": Object {
          "Fn::Join": Array [
            "",
            Array [
              "s3://",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/313bcae943a4b6c0cb79448b365008a0474eef65eca6898ee07874e5de502790.graphql",
            ],
          ],
        },
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "Todo": Object {
      "DependsOn": Array [
        "GraphQLAPITransformerSchema3CB2AE18",
      ],
      "Properties": Object {
        "Parameters": Object {
          "DynamoDBBillingMode": Object {
            "Ref": "DynamoDBBillingMode",
          },
          "DynamoDBEnablePointInTimeRecovery": Object {
            "Ref": "DynamoDBEnablePointInTimeRecovery",
          },
          "DynamoDBEnableServerSideEncryption": Object {
            "Ref": "DynamoDBEnableServerSideEncryption",
          },
          "DynamoDBModelTableReadIOPS": Object {
            "Ref": "DynamoDBModelTableReadIOPS",
          },
          "DynamoDBModelTableWriteIOPS": Object {
            "Ref": "DynamoDBModelTableWriteIOPS",
          },
          "referencetotransformerrootstackGraphQLAPI20497F53ApiId": Object {
            "Fn::GetAtt": Array [
              "GraphQLAPI",
              "ApiId",
            ],
          },
          "referencetotransformerrootstackGraphQLAPINONEDS2BA9D1C8Name": Object {
            "Fn::GetAtt": Array [
              "GraphQLAPINONEDS95A13CF0",
              "Name",
            ],
          },
          "referencetotransformerrootstackS3DeploymentBucket7592718ARef": "cdk-\${Qualifier}-assets-\${AWS::AccountId}-\${AWS::Region}",
          "referencetotransformerrootstackS3DeploymentRootKeyA71EA735Ref": "CdkBootstrap-\${Qualifier}-FileAssetKeyArn",
          "referencetotransformerrootstackenv10C5A902Ref": "NONE",
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/53d71e369f35572a33003a46a7dfb3e4a13cef81936b129fef00c0a7f03ccc02.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
